### PR TITLE
Dev 2386 buckets in axis

### PR DESCRIPTION
**Pull request by mehmetscoskun**

### CODE REVIEW

File: src/expressions/util/createDescendantGenerator.ts [hunk 1]
```diff
@@ -11,28 +11,30 @@ import { DONE_TOKEN, IterationHint, ready } from './iterators';
 
 function findDeepestLastDescendant(
 	node: ConcreteNode,
-	domFacade: IWrappingDomFacade
+	domFacade: IWrappingDomFacade,
+	bucket
```
Comment (JosVerburg) [inline src/expressions/util/createDescendantGenerator.ts]: Please use typing (at least for new code) `bucket: string`

File: src/expressions/util/createDescendantGenerator.ts [hunk 2]
```diff
@@ -11,28 +11,30 @@ import { DONE_TOKEN, IterationHint, ready } from './iterators';
...
 	}
 	return parentNode;
 }
 
 export default function createDescendantGenerator(
 	domFacade: IWrappingDomFacade,
 	node: ConcreteNode,
-	returnInReverse = false
+	returnInReverse = false,
+	bucket
```
Comment (JosVerburg) [inline src/expressions/util/createDescendantGenerator.ts]: Please use typing (at least for new code) `bucket: string`

File: src/types/Types.ts [hunk 3]
```diff
@@ -52,9 +52,9 @@ export type Document = Node & {
  */
 export type Element = Node & {
 	localName: string;
-	namespaceURI: string;
+	namespaceURI: string | null;
```
Comment (JosVerburg) [inline src/types/Types.ts]: This seems to be part of issue #122. Do we need this in this PR?
Let's either:
- Revert this change and fix it for `Attr` and `Element` in another PR.
- Also make the fix for `Attr` in this PR. (In this case we should also notice in the comment something like "This fixes #122")

File: test/specs/parsing/axes/AncestorAxis.tests.ts [hunk 4]
```diff
@@ -4,7 +4,9 @@ import jsonMlMapper from 'test-helpers/jsonMlMapper';
 
 import {
 	evaluateXPathToNodes,
-	evaluateXPathToStrings
+	evaluateXPathToStrings,
+	IDomFacade,
```
Comment (JosVerburg) [inline test/specs/parsing/axes/AncestorAxis.tests.ts]: Named imports must be alphabetized.tslint(ordered-imports)

File: test/specs/parsing/axes/AncestorAxis.tests.ts [hunk 5]
```diff
@@ -28,6 +30,26 @@ describe('ancestor', () => {
...
+		jsonMlMapper.parse([
+			'parentElement',
+			['childElement']
+		], documentNode);
+
+		const parentNode = documentNode.firstChild;
+		const childNode = documentNode.firstChild.firstChild;
+
+		const testDomFacade: IDomFacade = {
+			getParentNode: (node, bucket) => {
```
Comment (JosVerburg) [inline test/specs/parsing/axes/AncestorAxis.tests.ts]: `getParentNode: (node: slimdom.Node, bucket: string): slimdom.Node | null => {`

File: test/specs/parsing/axes/AncestorAxis.tests.ts [hunk 6]
```diff
@@ -28,6 +30,26 @@ describe('ancestor', () => {
...
+				chai.assert.notEqual(bucket, null, 'There must be a bucket passed!');
```
Comment (JosVerburg) [inline test/specs/parsing/axes/AncestorAxis.tests.ts]: We can use a specific null check `chai.assert.isNotNull(bucket, 'There must be a bucket passed!');`

File: test/specs/parsing/axes/AncestorAxis.tests.ts [hunk 7]
```diff
@@ -28,6 +30,26 @@ describe('ancestor', () => {
...
+				return getBucketsForNode(node.parentNode).includes(bucket) ? node.parentNode : null;
```
Comment (JosVerburg) [inline test/specs/parsing/axes/AncestorAxis.tests.ts]: Seems like a somewhat creative way to check the right buckets. Lets use `chai.assert.include(getBucketsForNode(node.parentNode), bucket);` and just return `null`.

File: test/specs/parsing/axes/AncestorAxis.tests.ts [hunk 8]
```diff
@@ -28,6 +30,26 @@ describe('ancestor', () => {
...
+			}
+		} as any;
+
+		const results = evaluateXPathToNodes('ancestor::parentElement', childNode, testDomFacade);
```
Comment (JosVerburg) [inline test/specs/parsing/axes/AncestorAxis.tests.ts]: When we do the check I described in the comment above, there will be no need left to check the result of xpath.
So this would than just be `evaluateXPathToNodes('ancestor::parentElement', childNode, testDomFacade);`.

File: test/specs/parsing/axes/AncestorAxis.tests.ts [hunk 9]
```diff
@@ -28,6 +30,26 @@ describe('ancestor', () => {
 		], documentNode);
 		chai.assert.deepEqual(evaluateXPathToStrings('ancestor-or-self::*/position()', documentNode.documentElement.firstChild), ['1', '2']);
 	});
+
+	it('passes buckets for ancestor', () => {
```
Comment (JosVerburg) [inline test/specs/parsing/axes/AncestorAxis.tests.ts]: Comments on this test file also apply on the others. I.e. typing, tslint, explicit tests.

File: test/specs/parsing/axes/AncestorAxis.tests.ts (hunk 7 above)
Comment (mehmetscoskun) [reply test/specs/parsing/axes/AncestorAxis.tests.ts]: The bucket comes from the selector, that's why `getBucketsFromNode(node.parentNode)` might not be include the bucket. for example, when parentNode is document node.

File: src/types/Types.ts (hunk 3 above)
Comment (DrRataplan) [reply src/types/Types.ts]: Let's fix the other one(s) too and just fix it at the same type

File: test/specs/parsing/axes/AncestorAxis.tests.ts [hunk 10]
```diff
@@ -14,50 +17,89 @@ beforeEach(() => {
...
+	});
+
+	it('passes buckets for ancestor', () => {
+		jsonMlMapper.parse(['parentElement', ['childElement']], documentNode);
+
+		const childNode = documentNode.firstChild.firstChild;
+
+		const testDomFacade: IDomFacade = {
+			getParentNode: (node: slimdom.Node, bucket: string | null) => {
+				chai.assert.include(getBucketsForNode(node.parentNode), bucket);
```
Comment (DrRataplan) [inline test/specs/parsing/axes/AncestorAxis.tests.ts]: I agree with this test. I like how it does not explicitly depend on the format of a bucket, but just tests whether it matches.

File: test/specs/parsing/axes/DescendantAxis.tests.ts [hunk 11]
```diff
@@ -14,44 +12,68 @@ beforeEach(() => {
...
+
+		const childNode = documentNode.firstChild.firstChild;
+
+		const testDomFacade: IDomFacade = {
+			getFirstChild: (node: slimdom.Node, bucket: string|null) => {
+				chai.assert.include(getBucketsForNode(node.childNodes[0]), bucket, 'It includes bucket');
+				return null;
+			},
+			getNextSibling: (node: slimdom.Node, bucket: string|null) => {
+				return null;
```
Comment (DrRataplan) [inline test/specs/parsing/axes/DescendantAxis.tests.ts]: Could you also add the assert here?

File: test/specs/parsing/axes/DescendantAxis.tests.ts [hunk 12]
```diff
@@ -14,44 +12,68 @@ beforeEach(() => {
...
+
+	it('passes buckets for descendant', () => {
+		jsonMlMapper.parse(['parentElement', ['childElement']], documentNode);
+
+		const childNode = documentNode.firstChild.firstChild;
+
+		const testDomFacade: IDomFacade = {
+			getFirstChild: (node: slimdom.Node, bucket: string|null) => {
+				chai.assert.include(getBucketsForNode(node.childNodes[0]), bucket, 'It includes bucket');
+				return null;
```
Comment (DrRataplan) [inline test/specs/parsing/axes/DescendantAxis.tests.ts]: If you return a childnode here, we will also execute the nextsibling method

File: test/specs/parsing/axes/FollowingAxis.tests.ts [hunk 13]
```diff
@@ -98,6 +104,34 @@ return map{
...
+
+		const firstChildNode = documentNode.firstChild.firstChild;
+
+		const testDomFacade: IDomFacade = {
+			getFirstChild: (node: slimdom.Node, bucket: string|null) => {
+				return null;
+			},
+			getNextSibling: (node: slimdom.Node, bucket: string|null) => {
+				chai.assert.include(getBucketsForNode(node.nextSibling), bucket, 'It includes bucket');
+				return null;
```
Comment (DrRataplan) [inline test/specs/parsing/axes/FollowingAxis.tests.ts]: Let's return something here, so that we can also assert the other two parts

File: test/specs/parsing/axes/FollowingSiblingAxis.tests.ts [hunk 14]
```diff
@@ -13,21 +11,59 @@ beforeEach(() => {
...
+		);
+
+		const firstChildNode = documentNode.firstChild.firstChild;
+
+		const testDomFacade: IDomFacade = {
+			getFirstChild: (node: slimdom.Node, bucket: string|null) => {
+				return null;
+			},
+			getNextSibling: (node: slimdom.Node, bucket: string|null) => {
+				chai.assert.include(getBucketsForNode(node.nextSibling), bucket, 'It includes bucket');
```
Comment (DrRataplan) [inline test/specs/parsing/axes/FollowingSiblingAxis.tests.ts]: Same, return, assert other two parts

File: test/specs/parsing/axes/PrecedingAxis.tests.ts [hunk 15]
```diff
@@ -122,6 +124,34 @@ return map{
...
+		const testDomFacade: IDomFacade = {
+			getLastChild: (node: slimdom.Node, bucket: string|null) => {
+				return null;
+			},
+			getParentNode: (node: slimdom.Node, bucket: string|null) => {
+				return null;
+			},
+			getPreviousSibling: (node: slimdom.Node, bucket: string|null) => {
+				chai.assert.include(getBucketsForNode(node.previousSibling), bucket, 'It includes bucket');
+				return null;
```
Comment (DrRataplan) [inline test/specs/parsing/axes/PrecedingAxis.tests.ts]: same, return node, assert two more

File: test/specs/parsing/axes/PrecedingSibling.tests.ts [hunk 16]
```diff
@@ -13,21 +11,59 @@ beforeEach(() => {
...
+
+		const testDomFacade: IDomFacade = {
+			getLastChild: (node: slimdom.Node, bucket: string|null) => {
+				return null;
+			},
+			getParentNode: (node: slimdom.Node, bucket: string|null) => {
+				return null;
+			},
+			getPreviousSibling: (node: slimdom.Node, bucket: string|null) => {
+				chai.assert.include(getBucketsForNode(node.previousSibling), bucket, 'It includes bucket');
```
Comment (DrRataplan) [inline test/specs/parsing/axes/PrecedingSibling.tests.ts]: same. return. assert

File: test/specs/parsing/axes/DescendantAxis.tests.ts [hunk 17]
```diff
@@ -14,44 +12,69 @@ beforeEach(() => {
...
+	});
+
+	it('passes buckets for descendant', () => {
+		jsonMlMapper.parse(['parentElement', ['childElement']], documentNode);
+
+		const childNode = documentNode.firstChild.firstChild;
+
+		const testDomFacade: IDomFacade = {
+			getFirstChild: (node: slimdom.Node, bucket: string|null) => {
+				chai.assert.isNotNull(bucket, 'It has bucket');
```
Comment (DrRataplan) [inline test/specs/parsing/axes/DescendantAxis.tests.ts]: We can make this test whether it passes the bucket for 'childElement' (`getBucketsForSelector('self::childElement')`)

File: src/domFacade/IDomFacade.ts [hunk 18]
```diff
@@ -6,13 +6,13 @@ import { Attr, CharacterData, Element, Node } from '../types/Types';
  * @public
  */
 export default interface IDomFacade {
-	getAllAttributes(node: Element): Attr[];
+	getAllAttributes(node: Element, bucket: string | null): Attr[];
```
Comment (DrRataplan) [inline src/domFacade/IDomFacade.ts]: We could add some documentation here:

```
/**
 * Get all attributes of this element
 * @param  node
 * @param  bucket  The bucket that matches the attribute that will be used. This bucket can be used to narrow down which attributes should be retrieved.
 */

File: src/domFacade/IDomFacade.ts [hunk 19]
```diff
@@ -6,13 +6,13 @@ import { Attr, CharacterData, Element, Node } from '../types/Types';
...
+	getChildNodes(node: Node, bucket: string | null): Node[];
 	getData(node: Attr | CharacterData): string;
-	getFirstChild(node: Node): Node | null;
-	getLastChild(node: Node): Node | null;
-	getNextSibling(node: Node): Node | null;
-	getParentNode(node: Node): Node | null;
-	getPreviousSibling(node: Node): Node | null;
+	getFirstChild(node: Node, bucket: string | null): Node | null;
+	getLastChild(node: Node, bucket: string | null): Node | null;
+	getNextSibling(node: Node, bucket: string | null): Node | null;
```
Comment (DrRataplan) [inline src/domFacade/IDomFacade.ts]:
```
/**
 * Get the next sibling of this node
 * @param  node
 * @param  bucket  The bucket that matches the nextSibling that is requested. An implementation of IDomFacade is free to interpret this to skip returning nodes that do not match the bucket, or use this information to its advantage.
 */
```
